packages/realm/src/schema/types.ts

```diff
@@ -71,30 +101,221 @@ export type CanonicalObjectSchemaProperty = {
+ *   indexed. In such cases, they cannot be made non-indexed.
+ *
+ * @see {@link PropertySchemaStrict} for a precise type definition of the requirements
+ * with the allowed combinations. This type is less strict in order to provide a more
+ * user-friendly option due to misleading TypeScript error messages when working with
+ * the strict type. This type is currently recommended for that reason, but the strict
+ * type is provided as guidance. (Exact errors will always be shown when creating a
+ * {@link Realm} instance if the schema is invalid.)
+ */
+export type PropertySchema = {
+  /**
+   * The type of the property.
+   */
+  type: PropertyTypeName;
+  /**
+   * The type of the elements in the collection if `type` is a {@link CollectionPropertyTypeName},
+   * or the specific Realm object type if `type` is a {@link RelationshipPropertyTypeName}.
+   */
+  objectType?: PrimitivePropertyTypeName | UserTypeName;
+  /**
+   * The name of the property of the object specified in `objectType` that creates this link.
+   */
+  property?: string;
+  /**
+   * Whether to allow `null` or `undefined` to be assigned to the property; or in the
+   * case of a collection, to be assigned to its elements. (Realm object types in lists
+   * and sets cannot be optional.)
+   *
+   * Default value: See the requirements in the documentation for this type.
+   */
+  optional?: boolean;
+  /**
+   * Whether the property should be indexed. (Only supported if `type` is `"string"`,
+   * `"int"`, or `"bool"`).
+   *
+   * Default value: `false` if the property is not a primary key, otherwise `true`.
+   */
+  indexed?: boolean;
+  /**
+   * The name to be persisted in the Realm file if it differs from the already-defined
+   * JavaScript/TypeScript (JS/TS) property name. This is useful for allowing different
+   * naming conventions than what is persisted in the Realm file. Reading and writing
+   * properties must be done using the JS/TS name, but queries can use either the JS/TS
+   * name or the persisted name.
+   */
+  mapTo?: string;
+  /**
+   * The default value that the property will be set to when created.
+   */
+  default?: unknown;
+};
+
+/**
+ * Keys used in the property schema that are common among all variations of {@link PropertySchemaStrict}.
+ */
+type PropertySchemaCommon = {
+  indexed?: boolean;
+  mapTo?: string;
+  default?: unknown;
+};
+
+/**
+ * The strict schema for specifying the type of a specific Realm object property.
+ *
+ * Unlike the less strict {@link PropertySchema}, this type precisely defines the type
+ * requirements and their allowed combinations; however, TypeScript error messages tend
+ * to be more misleading. {@link PropertySchema} is recommended for that reason, but the
+ * strict type is provided as guidance.
+ *
+ * @see {@link PropertySchema} for a textual explanation of the requirements defined here,
+ *   as well as documentation for each property.
+ */
+export type PropertySchemaStrict = PropertySchemaCommon &
```

Comment:
Is this type to be used by the developer using Realm? Or is it simply here for documentation purposes?

Reply:
Users would most likely not explicitly define a `myKey: undefined`, but was thinking more if this strict type should align with the rules of the parser (allowing `undefined`) or just not allow the field at all. Mathias, so are you saying you would vote for keeping the fields?